ADR/0040-worspace-design-proposal.md

```diff
@@ -0,0 +1,172 @@
+# 40. Workspace Abstraction
+
+Date: 2024-9-26
+
+## Status
+
+WiP
+
+## Context
+
+Workspaces are being handled differently today in our community and in our downstream deployments. 
+
+### Downstream
+
+We have a workspace CRD defined with a controller that interfaces with KubeSaw as the NSaaS implementation.
+
+Kubesaw provides multiple features but mainly namespace provisioning, multi-cluster placement algorithms, syncronization of tiers (Quotas, Roles, RoleBindings and ConfigMaps) and user lifecycle management.
+
+Workspace CRD serves as an abstraction to be able to make changes to a workspace or create new workspaces for a specific users. 
+
+This abstraction also enable Konflux to implement other features in the future like for example limmiting the amount of workspaces per user, allowing workspace delete self-service and adding workspace delete protection.
+
+
+### Upstream
```

Comment:
I would recommend making one step back and looking at the requirements and expectation you have for upstream. What do you want to provide as part of the upstream project?
Is it only a secured build system?
Do you want to also provide namespace management tooling?
How about personas and their ability to manage namespaces and other resources in the cluster? Is it something you want to provide in Konflux as well or not?

My point is, do you want to focus on implementing the secured pipeline only, or do you want to try to solve anything related to multi-tenancy as well? Or would you rather delegate this to the setup that is used by the customer/consumer? 
The environments where Konflux will be used may have tens of ways of managing namespaces and access to them (KubeSaw, ArgoCD, Kyverno, kubectl, RHDH, OpenShift projects, custome in-house implementation, ...). I wouldn't try to limit it to only a small set of specific supported tools if the project should be really adopted in an already existing environments. Because keep in mind that there are running clusters that already have many namespaces created, how should be Konflux integrated in such an environment?
In addition to that, each company may have completely different set of policies for managing access to clusters. There can be small start-ups that doesn't care and everyone has cluster-admin permissions, but there are can be companies where devs can only view stuff, or don't have any access to the cluster at all.

And just to be clear, the before-mentioned text is about requirements for "upstream" only. 
The requirements for the internal running Konflux instance is completely different topic. Red Hat is just one adopter of Konflux project that has its own specific requirements on how the namespaces and access is managed there.